internal/cluster/connection.go

```diff
@@ -0,0 +1,256 @@
+	startTime                 int64
+	logger                    ilogger.Logger
+}
+
+func (c *Connection) ConnectionID() int64 {
+	return c.connectionID
+}
+
+func (c *Connection) start(clusterCfg *pubcluster.Config, addr pubcluster.Address) error {
+	if socket, err := c.createSocket(clusterCfg, addr); err != nil {
+		return err
+	} else {
+		c.socket = socket
+		c.lastWrite.Store(time.Time{})
+		c.closedTime.Store(time.Time{})
+		c.startTime = timeutil.GetCurrentTimeInMilliSeconds()
+		c.lastRead.Store(time.Now())
+		if err := c.sendProtocolStarter(); err != nil {
+			c.socket.Close()
+			c.socket = nil
+			return err
+		}
+		go c.socketReadLoop()
+		go c.socketWriteLoop()
+		return nil
+	}
+}
+
+func (c *Connection) sendProtocolStarter() error {
+	_, err := c.socket.Write([]byte(protocolStarter))
+	return err
+}
+
+func (c *Connection) createSocket(networkCfg *pubcluster.Config, address pubcluster.Address) (net.Conn, error) {
+	conTimeout := timeutil.GetPositiveDurationOrMax(networkCfg.ConnectionTimeout)
+	socket, err := c.dialToAddressWithTimeout(address, conTimeout)
+	if err != nil {
+		return nil, err
+	}
+	//if networkCfg.SSLConfig().Enabled() {
+	//	socket, err = c.openTLSConnection(networkCfg.SSLConfig(), socket)
+	//}
+	return socket, err
+}
+
+func (c *Connection) dialToAddressWithTimeout(addr pubcluster.Address, conTimeout time.Duration) (*net.TCPConn, error) {
+	if conn, err := net.DialTimeout("tcp", addr.String(), conTimeout); err != nil {
+		return nil, err
+	} else {
+		tcpConn := conn.(*net.TCPConn)
+		if err = tcpConn.SetNoDelay(true); err != nil {
+			panic(fmt.Errorf("error setting tcp no delay: %w", err))
+		}
+		if err = tcpConn.SetReadBuffer(bufferSize); err != nil {
+			panic(fmt.Errorf("error setting read buffer: %w", err))
+		}
+		if err = tcpConn.SetWriteBuffer(bufferSize); err != nil {
+			panic(fmt.Errorf("error setting write buffer: %w", err))
+		}
+		return tcpConn, nil
+	}
+}
+
+func (c *Connection) isAlive() bool {
+	return atomic.LoadInt32(&c.status) == open
+}
+
+func (c *Connection) socketWriteLoop() {
+	for {
+		select {
+		case request, ok := <-c.pending:
+			if !ok {
+				return
+			}
+			if err := c.write(request); err != nil {
+				c.logger.Errorf("write error: %w", err)
+				// XXX: create a new client message?
+				request.Err = err
+				c.responseCh <- request
+				c.close(err)
+			} else {
+				c.lastWrite.Store(time.Now())
+			}
+		case <-c.doneCh:
+			return
+		}
+	}
+}
+
+func (c *Connection) socketReadLoop() {
+	var err error
+	var n int
+	buf := make([]byte, bufferSize)
+	clientMessageReader := newClientMessageReader()
+	for {
+		//c.socket.SetReadDeadline(time.Now().Add(1 * time.Second))
+		n, err = c.socket.Read(buf)
+		if atomic.LoadInt32(&c.status) != open {
+			break
+		}
+		if err != nil {
+			if c.isTimeoutError(err) {
+				continue
+			}
+			c.logger.Errorf("read error: %w", err)
+			break
+		}
+		if n == 0 {
+			continue
+		}
+		clientMessageReader.Append(buf[:n])
```

Comment:
nit: memory copying could be avoided here in situations when the input buffer (byte slice) contains all parts of a response message. In that case, we could reslice the buffer into frames while reading it.

Reply:
We should definitely do something with the current `socketReadLoop()` (and underlying `clientMessageReader.resetBuffer()`) behavior, since it shows as the Top 1 source of allocations in `pprof` memory profile.

![Screenshot from 2021-04-26 16-19-33](https://user-images.githubusercontent.com/37772591/116089156-3abcc400-a6ab-11eb-93ac-42e9944d8a51.png)